File: src/crates/core/src/state/state.rs
use shared::state::renderer::{WindowSize, Position};
pub struct State {
    pub audio_active: bool,
    pub speed: f64,
    pub window_size: WindowSize,
    pub ball_position: Position,
}

//Initial state
impl State {
    pub fn new(window_width: u32, window_height: u32) -> Self {
        Self {
            audio_active: true, 
            speed: 0.5,
            window_size: WindowSize { width: window_width, height: window_height},
            ball_position: Position { x: 0.0, y: 0.0}
        }
    }
}

File: typescript/state/state.ts
import { interpret, Machine, spawn, assign } from "xstate";
import {state_transition_event} from "@events/events";
import {debug_settings} from "@config/config";

const machine = Machine({
    id: "main",
    initial: "init",
    context: {
        speed: .5,
        window_width: 0,
        window_height: 0,
        audio_active: true 
    },
    states: {
        init: {
            on: { 
                READY: [
                    {target: "start_loading", cond: () => debug_settings.skipStart},
                    {target: "waiting"},
                ]
            }
        },
        waiting: {
            on: {
                START: "start_loading"
            }
        },

        start_loading: {
            onEntry: () => state_transition_event("start"),
            on: {
                ASSETS_LOADED: "running"
            }
        },

        running: {
            on: {
                TOGGLE_AUDIO: {
                    actions: assign({
                        audio_active: (ctx:any) => !ctx.audio_active
                    }) as any
                }
            }
        }
    }
});

let _service;

export const get_service = () => {
    if(!_service) {
        _service = 
            interpret(machine).start();
    }

    return _service;
}

export const get_state = () => get_service().state;